engine/src/main/java/org/hibernate/search/engine/search/predicate/dsl/OrPredicateOptionsStep.java

```diff
@@ -0,0 +1,46 @@
+/*
+ * Hibernate Search, full-text search for your domain model
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+package org.hibernate.search.engine.search.predicate.dsl;
+
+import java.util.function.Consumer;
+
+import org.hibernate.search.engine.search.predicate.SearchPredicate;
+
+/**
+ * The final step in an "or" predicate definition, where optional parameters can be set.
+ *
+ * @param <S> The "self" type (the actual exposed type of this step).
+ */
+public interface OrPredicateOptionsStep<S extends OrPredicateOptionsStep<?>>
+		extends PredicateScoreStep<S>, PredicateFinalStep {
+	/**
+	 * Delegates setting clauses and options to a given consumer.
+	 * <p>
+	 * Best used with lambda expressions.
+	 *
+	 * @param contributor A consumer that will add clauses and options to the collector that it consumes.
+	 * Should generally be a lambda expression.
+	 * @return {@code this}, for method chaining.
+	 */
+	S with(Consumer<? super S> contributor);
```

Comment:
This will allow things such as:

```
f.and().with( c -> c.toPredicate() ).add( ... )
```

Which makes no sense, because the lambda has no business calling `toPredicate` to begin with. So we probably shouldn't expose it to the consumer?

I suggest you take more inspiration from the interface hierarchy of `org.hibernate.search.engine.search.predicate.dsl.BooleanPredicateOptionsCollector`: it is complicated, but that complexity **is** solving problems, and you're facing very much the same problems.

Reply:
I thought this was requested in the PR (second set of patterns in "Builder-style syntaxes may be useful as well") and I took inspiration from `BooleanPredicateOptionsCollector`...  I'll have a second look!

Reply:
The problem is *only* that the consumer is able to call `toPredicate`; calling `add` like in the JIRA issue, or even `boost()`, in the consumer is fine. But `toPredicate` is a terminal operation, so it doesn't have its place in the `with` call (because the `with` call will return a non-terminal step and allow more changes).

Reply:
Okay...  I think I got your point!
I reworked the class hierarchy so that it gets more similar to the `BooleanPredicateOptionsCollector` one and `toPredicate()` is now no longer reachable within `with(...)`.
I think I now get better the point of the `GenericBooleanPredicateClausesStep` interface that didn't seem to add much in term of... interface and seemed to mostly repeat the super interface: it actually refines it in order to distinguish which parameterized type is "below" the current level, and what is above it, and why the latter is only used in the `Consumer`: the point is to hide the `PredicateFinalStep` interface there.
Quite tricky but clever!

(I never needed such restrictions even in the worst DSL I ever wrote for building some kind of god object and clarify which settings must be set at instantiation time and which one may be changed later on...)

I end up with 3 interfaces and 2 classes (* 2 for `and` and `or`), as for the beef of `BooleanPredicateOptionsCollector` and couldn't manage to have less...  :'-(
But I do agree that this is worth it, especially as `toPredicate()` is "silently terminal", as it will cause the predicate to be built and all other modifications will not reset the "built state".

I will have to rework that to address https://github.com/hibernate/hibernate-search/pull/3152#discussion_r934413732, though...  :'-(

And now I'm wondering whether we should have regression tests for methods that should not be accessible in some places (=> code that should not compile)...  ;-)
Looks like Javassist also supports working on (simple) Java source code, or the javac API could be used (maybe via https://github.com/OpenHFT/Java-Runtime-Compiler), but that would be for another PR, maybe!  ;-)

Reply:
> Okay... I think I got your point!

You did, thanks for bearing with me :)

> I end up with 3 interfaces and 2 classes (* 2 for `and` and `or`), as for the beef of `BooleanPredicateOptionsCollector` and couldn't manage to have less... :'-(

That seems fine to me. You *might* be able to take it down to 3*2 interfaces and 2*1 classes by merging the `and`/`or` implementations, with an instance member representing the operator you're working with (the `BooleanOperator` enum will come in handy). But I don't think that's absolutely necessary.

> And now I'm wondering whether we should have regression tests for methods that should not be accessible in some places (=> code that should not compile)... ;-)

That's an interesting idea, but I'm not sure that would be worth the effort: there are so many things that shouldn't compile!

>  Looks like Javassist also supports working on (simple) Java source code, or the javac API could be used (maybe via https://github.com/OpenHFT/Java-Runtime-Compiler), but that would be for another PR, maybe! ;-)

Javassist is a no-no; Hibernate projects have been moving away from it due to significant problems with recent versions of Java; from what I remember there was fundamental problems with the approach Javassist took that prevented it from working perfectly with recent versions of Java.

There are other tools for that (like the one you mentioned), but overall I think a better approach would be automatic API diffs. We've explored that in the past, but never went as far as making the build fail as soon as there's a backward-incompatible change, and certainly not as soon as there's any change, compatible or not. That might be an idea, though :)

Reply:
> thanks for bearing with me :)

I actually totally appreciate the rigor!

> You might be able to take it down to 3*2 interfaces and 2*1 classes by merging the and/or implementations, with an instance member representing the operator you're working with [...]. But I don't think that's absolutely necessary.

I really think there's too much code that is 99% similar to keep it that way: the only added value I see here is the explicit interface/class names (with `And` / `Or` in them), and that doesn't seem to be worth the trouble.  I'd say that variations around something operator-agnostic like `SimplePredicate` should be clear enough.  Don't you think?
\+ adding support for `mustNot` will make things even worse.

> the `BooleanOperator` enum will come in handy

Hmmm...  I was more thinking about defining my own local `enum` (esp. as `BooleanOperator` doesn't support `mustNot`) in order to attach the proper `builder` method to invoke (`must`, `should`, etc.).  Or did I miss something ?

Reply:
> I'd say that variations around something operator-agnostic like SimplePredicate should be clear enough. Don't you think?

So first, I think `Simple` is a very bad name as it conveys little meaning :) `BooleanJunctionPredicateOptionsStep` might cut it, though I must admit I don't find that name very good either: it's long and not very easy to understand.

If you can find a good name, then sure, you can go a step further and expose a single set of interfaces.

> Hmmm... I was more thinking about defining my own local enum (esp. as BooleanOperator doesn't support mustNot) in order to attach the proper builder method to invoke (must, should, etc.). Or did I miss something ?

When you mention `mustNot`, you are talking about the `not` operator we'll ultimately add?

I would have expected that to be an unary operator, without the builder API. But if you think passing multiple predicates to `not` makes sense, ok, why not. Local enum it is.

Reply:
> > I'd say that variations around something operator-agnostic like SimplePredicate should be clear enough. Don't you think?
> 
> So first, I think `Simple` is a very bad name as it conveys little meaning :) `BooleanJunctionPredicateOptionsStep` might cut it, though I must admit I don't find that name very good either: it's long and not very easy to understand.
> 
> If you can find a good name, then sure, you can go a step further and expose a single set of interfaces.

Seeing your code, maybe `ScoredPredicateOptionsStep` + `CompositePredicateClausesStep` would fit?

Reply:
> So first, I think Simple is a very bad name as it conveys little meaning :)

Not a fan either.
I reasoned from the documentation, where I think we should first introduce these leaner predicates, which are meant to address many simple use cases, and then move on to more complex, but more powerful, `bool`()-based predicates.
BTW, WDYTAT?

> When you mention `mustNot`, you are talking about the `not` operator we'll ultimately add?

Well, I thought about adding support for `bool().mustNot( predicate1 ).mustNot( predicate2 )...` (and all related syntaxes), with something like `not( predicate1, predicate2, ... )`, which is actually different from what you've described in https://hibernate.atlassian.net/browse/HSEARCH-4645.
Does my proposition make sense, regardless of HSEARCH-4645?
If so, this could be very easy to support: the most complex part would be to recycle existing `mustNot` tests.  ;-)

> Seeing your code, maybe `ScoredPredicateOptionsStep` + `CompositePredicateClausesStep` would fit?

Why not...  We'll miss the opposition between "simple" and "more complex" predicates and this will make less clear that these `...Options...` and `...Clauses...` classes are related but who said that the class hierarchy should be redundantly encoded in class names?  ;-)

Reply:
>   I reasoned from the documentation, where I think we should first introduce these leaner predicates, which are meant to address many simple use cases, and then move on to more complex, but more powerful, bool()-based predicates.
BTW, WDYTAT?

I'm not sure what you mean by "moving on to ...". You want to make changes to `.bool()`?

>  Does my proposition make sense, regardless of HSEARCH-4645?

It seems very similar to HSEARCH-4645 to me. If you want to do more than what's suggested in that ticket, maybe that's worth discussing on the ticket?

In any case, you've already worked much more than expected on this ticket, I don't think it would be wise to address more problems in the same pull request :)

> Why not... We'll miss the opposition between "simple" and "more complex" predicates and this will make less clear that these ...Options... and ...Clauses... classes are related but who said that the class hierarchy should be redundantly encoded in class names? ;-)

Right, there's a balance to be found between expressiveness and brevity. We can't always put all the information in the name :)

> What about just replacing Simple by Composite (even though more complex predicates are even more composite)? ;-)

You mean `CompositePredicateOptionsStep` + `CompositePredicateClausesStep`? That means we won't be able to use `CompositePredicateOptionsStep` for... non-composite predicates, but since we don't need to at the moment anyway, sure, that's probably best.

Reply:
> I'm not sure what you mean by "moving on to ...". You want to make changes to `.bool()`?

I was referring to the documentation: I was thinking that we should introduce the new simple predicate combinator _before_ [9.2.11. bool: combine predicates (or/and/…​)](https://docs.jboss.org/hibernate/stable/search/reference/en-US/html_single/#search-dsl-predicate-boolean), and then present `bool()` as a way to combine predicates in a more powerful way, albeit with a slightly more complex syntax.

> In any case, you've already worked much more than expected on this ticket

Well, not that much in term of "CPU" time, but the wall time was really awful...
(But, you know, it's summer and...)

> I don't think it would be wise to address more problems in the same pull request :)

Okay, let's keep that in mind if you ever think it's worth it.

> `...Composite...`

I really think we should find a way to oppose `bool()`-based predicates and these new ones, that only support a single operator, at least at top-level (nesting "simple" predicates remains possible).
Looks to me like the word we are looking for is something around "uniform".  ;-)
So what about keeping the existing names (and esp. the distinction between `Options` and `Clauses` steps) and just replacing `Simple` by `Uniform`?

Reply:
> > I'm not sure what you mean by "moving on to ...". You want to make changes to `.bool()`?
> 
> I was referring to the documentation: I was thinking that we should introduce the new simple predicate combinator _before_ [9.2.11. bool: combine predicates (or/and/…​)](https://docs.jboss.org/hibernate/stable/search/reference/en-US/html_single/#search-dsl-predicate-boolean), and then present `bool()` as a way to combine predicates in a more powerful way, albeit with a slightly more complex syntax.

Right; this makes sense.

> > `...Composite...`
> 
> I really think we should find a way to oppose `bool()`-based predicates and these new ones, that only support a single operator, at least at top-level (nesting "simple" predicates remains possible). Looks to me like the word we are looking for is something around "uniform". ;-) So what about keeping the existing names (and esp. the distinction between `Options` and `Clauses` steps) and just replacing `Simple` by `Uniform`?

I'm not sure. We're going into philosophical territory at this point, but let's give it a try...

`SimplePredicate*Step` works for `.and()`/`.or()`, but this name could arguably be applied to about any predicate (the `matchAll` predicate is quite "simple" after all), so it's not specific enough.

As you pointed out, `CompositePredicate*Step` suffers from a similar problem: it could apply as much to `.bool()` as to `.and()`/`.or()`.

`UniformPredicate*Step` may be clear to us while this discussion is still fresh, but I doubt it still will be two weeks from now. Besides, `and`/`or` are arguably not the only "uniform" predicates; who's to say the "match" predicate is not "uniform"? :)

`SimpleBooleanJunctionPredicate*Step` would reflect what we mean, more or less, but it's long.

`SimpleBooleanPredicate*Step` is more reasonable (still long, but I've done worse). It could potentially also apply to the `not` predicate (even though we might not expose the `*ClausesStep` for that one, but that's a separate discussion). WDYT?

Reply:
"Boolean" along with "Predicate" always sounds like a pleonasm to me, but I do understand the will to refer to the `bool()` construct, which is what is ultimately used.

So...  `*SimpleBooleanPredicate*Step*` it is?  ;-)

Reply:
> "Boolean" along with "Predicate" always sounds like a pleonasm to me, but I do understand the will to refer to the `bool()` construct, which is what is ultimately used.

Well we can go with `SimpleBooleanOperatorPredicate` if you want, but... we're back to a long name :)

> So... *SimpleBooleanPredicate*Step* it is? ;-)

Works for me, at least :) Feel free to use `SimpleBooleanOperatorPredicate` if you prefer that.

Reply:
Let's go for `SimpleBooleanOperatorPredicate`, then!
Gotta find a way to justify having an `Abstract...FactoryBuilderRepositoryManager`, and we'll be just fine!  ;-)